src/index/GraphFilter.h

```diff
@@ -0,0 +1,111 @@
+  GraphFilter& operator=(GraphFilter&&) noexcept = default;
+
+  // Keep all graphs.
+  static GraphFilter All();
+  // Only keep graphs in `whitelist`.
+  static GraphFilter Whitelist(ad_utility::HashSet<T> whitelist);
+  // Keep all graphs that are not `value`.
+  static GraphFilter Blacklist(T value);
+
+  // Transforms this `GraphFilter` into another one with a different template
+  // argument, by using `func` to transform the underlying values from `T` to a
+  // new type.
+  CPP_template(typename Func)(requires std::invocable<Func, const T&>)
+      GraphFilter<decltype(std::declval<Func>()(std::declval<T>()))> transform(
+          const Func& func) const {
+    using TransformedT = decltype(func(std::declval<T>()));
+    using ResultT = GraphFilter<TransformedT>;
+    return std::visit(ad_utility::OverloadCallOperator{
+                          [](const AllTag&) { return ResultT::All(); },
+                          [&func](const ad_utility::HashSet<T>& whitelist) {
+                            ad_utility::HashSet<TransformedT> result;
+                            for (const T& value : whitelist) {
+                              result.emplace(func(value));
+                            }
+                            return ResultT::Whitelist(std::move(result));
+                          },
+                          [&func](const T& blacklist) {
+                            return ResultT::Blacklist(func(blacklist));
+                          }},
+                      filter_);
+  }
+
+  // Return true iff the graph `graph´ is allowed by this filer.
+  bool isGraphAllowed(T graph) const;
+
+  // Return true iff all graphs are always allowed.
+  bool isWildcard() const;
+
+  // Make sure this filter is comparable.
+  bool operator==(const GraphFilter&) const = default;
+
+  // Describe this `GraphFilter` and write it to the output stream using
+  // `formatter` to format the individual graph values `T`.
+  CPP_template(typename Formatter)(
+      requires ad_utility::InvocableWithExactReturnType<
+          Formatter, std::string, const T&>) void format(std::ostream& os,
+                                                         const Formatter&
+                                                             formatter) const {
+    os << "GRAPHS: ";
+    std::visit(ad_utility::OverloadCallOperator{
+                   [&os](const AllTag&) { os << "ALL"; },
+                   [&os, &formatter](const ad_utility::HashSet<T>& whitelist) {
+                     os << "Whitelist "
+                        << absl::StrJoin(
+                               whitelist | ql::views::transform(formatter), " ")
+                        << std::endl;
+                   },
+                   [&os, &formatter](const T& blacklist) {
+                     os << "Blacklist " << std::invoke(formatter, blacklist)
+                        << std::endl;
+                   }},
+               filter_);
+  }
+};
```

Comment:
As string representations are "slow" anyway, you can make the argument an `absl::FunctionRef` and hide the implementation in the .cpp file.